src/action/install_date.rs

```diff
@@ -0,0 +1,284 @@
+            let parsed = DateTime::<Utc>::from(parsed);
+            assert_eq!(parsed, correct);
+        }
+    }
+}
+
+/// Obtains system install date (Linux version).
+///
+/// To obtain install date, it first checks the creation time of the root
+/// filesystem. Then, it searches for the installation logs and gets their
+/// modification date. It also checks for other files that were probably
+/// unchanged from the installation.
+///
+/// This function returns `None` in case of errors.
+#[cfg(target_os = "linux")]
+fn get_install_time() -> Option<SystemTime> {
+    // First, check the creation time of the root. This method works on
+    // Linux >= 4.11.
+    let time = fs::metadata("/").and_then(|metadata| metadata.created());
+    if let Ok(time) = time {
+        return Some(time);
+    }
+
+    // Then, try to detect filesystem creation time using dumpe2fs. This
+    // works only on ext2/3/4 filesystems.
+    let time = e2fs_utils::creation_time_from_dumpe2fs();
+    if let Some(time) = time {
+        return Some(time);
+    }
+
+    // Then, search for various files that were potentially modified only
+    // on installation:
+    // * /var/log/installer/: The location of installation logs for Debian-based
+    //   distributions.
+    // * /root/install.log: The location of installation logs for RHEL-based
+    //   distributions.
+    // * /etc/hostname: In most cases, we may assume that the hostname didn't
+    //   change after installation.
+    // * /lost+found: This method was used by Python version on GRR client, so
+    //   leaving it here.
+    const CANDIDATES: [&str; 5] = [
+        "/var/log/installer/syslog",
+        "/var/log/installer/status",
+        "/root/install.log",
+        "/etc/hostname",
+        "/lost+found",
+    ];
+    if let Some(time) = get_modified_time(CANDIDATES.iter()) {
+        return Some(time);
+    }
+
+    // We tried our best and all the methods have failed, so just give up.
+    None
+}
+
+/// Obtains system install date (OS X version).
+///
+/// This function searches for some files that were probably unchanged since
+/// the installation.
+///
+/// This function returns `None` in case of errors.
+#[cfg(target_os = "macos")]
+fn get_install_time() -> Option<SystemTime> {
+    // Here, we use the same way as Python version of GRR client does. We just
+    // check the modification time for some of the paths
+    const CANDIDATES: [&str; 3] = [
+        "/var/log/CDIS.custom",
+        "/var",
+        "/private",
+    ];
+    get_modified_time(CANDIDATES.iter())
+}
+
+/// Obtains system install date (Windows version).
+///
+/// To obtain install date, this function searches the corresponding value in
+/// Windows registry.
+///
+/// This function returns `None` in case of errors.
+#[cfg(target_os = "windows")]
+fn get_install_time() -> Option<SystemTime> {
+    use winreg::RegKey;
+
+    // Don't use winreg::enums::KEY_WOW64_64KEY since it breaks on Windows 2000
```

Comment:
Could you clarify what does it mean to "break" in this context? Also: missing dot at the end.

Reply:
I don't know, this comment was just copied from GRR together with the idea of the implementation :)

Reply:
Also, I'm not sure whether Rust support Windows 2000 or anyone is still using it. If not, this comment can be safely removed.